history/storage.py

```diff
@@ -116,69 +164,61 @@ def retrieve_response(self, spider, request):
 
         data = json.loads(data_string)
 
-        metadata         = data['metadata']
-        request_headers  = Headers(data['request_headers'])
-        request_body     = data['request_body']
+        metadata = data['metadata']
         response_headers = Headers(data['response_headers'])
-        response_body    = data['response_body']
-
-        if 'binary' in data and data['binary'] == True:
-            logger.debug('S3Storage: retrieved binary body')
-            response_body = base64.decode(response_body)
-
-        url      = metadata['response_url']
-        status   = metadata.get('status')
-
-        logger.debug('S3Storage: response headers {} '.format(response_headers))
-        Response = responsetypes.from_args(headers=response_headers, url=url, body=response_body)
-        logger.debug('S3Storage: response type {} '.format(Response))
-
-        return Response(url=url, headers=response_headers, status=status, body=response_body)
+        response_body = data['response_body']
+        
+        if data.get('binary', False):
+            logger.debug('retrieved binary body')
+            response_body = base64.b64decode(response_body.decode('utf8'))
+            encoding = {}
+        else:
+            encoding = {'encoding': 'utf8'}
+        url = str(metadata['response_url'])
+        status = metadata.get('status')
+        logger.debug('response headers {}'.format(response_headers))
+        Response = responsetypes.from_args(headers=response_headers, url=url)
+        logger.debug('response type {}'.format(Response))
+        return Response(url=url,
+                        headers=response_headers,
+                        status=status,
+                        body=response_body,
+                        **encoding)
 
     def store_response(self, spider, request, response):
         """
         Store the given response in the cache.
+
         """
-        logger.info('S3Storage: storing response for %s.' % request.url)
+        logger.info('storing response for %s.' % request.url)
         key = self._get_key(spider, request)
+        logger.info('path %s' % key)
```

Comment:
I'm not sure the `logger` usage here is production-compliant. Logs typically tell what the program does and here we only debug variable, without any context (and line 195 looks much more like a debug statement).
Although I recommended to use `logger.debug` while developing, variable print statement shouldn't end up in production (and open source) code shared among the team.